Use Point/Displacements when searching the area centered on a tile

ephphatha opened the pull request:
This addresses a todo comment, though not in the way that was originally intended as we no longer use Direction::Omni

AJenbo commented on Source/controls/plrctrls.cpp:
```diff
@@ -122,61 +122,80 @@ int GetDistanceRanged(Point destination)
 {
-	int mx = Players[MyPlayerId].position.future.x;
-	int my = Players[MyPlayerId].position.future.y;
+	// This search uses row-major order, compared to CheckThemeObj3 using col-major :(
```
Feel free to change this one

ephphatha commented:
To elaborate on the last commit, in vanilla Diablo/Hellfire opening a trapped chest fires the missile toward the player closest to the left/bottom of the trapped object. In the following screenshot the warrior would be targeted. With the early break, the rogue gets targeted instead. I thought it was a minor change to avoid looping unnecessarily in multiplayer games but it can be ignored pretty easily.

![image](https://user-images.githubusercontent.com/357684/139634177-0f42f268-a820-4748-94dc-a97e8d5b1213.png)

The other behavioural change relates to moving with a controller which was added in devX anyway, this means that the search prefers the item above/to the right of the player when they move between two items or interactable objects.

julealgon commented on Source/path.h:
```diff
@@ -5,14 +5,26 @@
+		    { -1,  1 }, { 0,  1 }, { 1,  1 },
+	    // clang-format on
+	}
+};
```
Can't this be done using a 3x3 rectangle?

I understand using individual points for more "complex" shapes (like quest entry points) since we don't have a general purpose `Region` abstraction, but for these simpler areas, it feels to me there is no reason to go with individual displacements in place of a `Rectangle` instance that is applied at tile space.

ephphatha replied:
Probably, I'll see what it looks like at some point over the weekend.

ephphatha commented:
@julealgon Took a fair bit of boilerplate but the last commit adds a wrapper to allow iterating over all points (tiles) covered by a rectangle. I added some constructors to make it easier to define a square centered on a point however the range class works with arbitrary rectangles.

Would definitely appreciate more eyes on the iterator since it's the first time I've attempted something which doesn't return an lvalue.

AJenbo commented on Source/engine/size.hpp:
```diff
@@ -6,6 +6,20 @@ struct Size {
+	{
+	}
+
+	explicit constexpr Size(int size)
+	    : width(size)
+	    , height(size)
+	{
+	}
```
Do we finally have a solution the the rectangle vs square problem of programming :o

ephphatha commented:
Given I've introduced (now fixed) an off by one error I'll add some tests for this iterator after work. It'll help document the ordering requirements needed for use during dungeon generation.

glebm commented on Source/engine/points_in_rectangle_range.hpp:
```diff
@@ -0,0 +1,195 @@
+#pragma once
+
+#include "point.hpp"
```
```suggestion
#include <iterator>

#include "point.hpp"
```

glebm commented on Source/engine/points_in_rectangle_range.hpp:
```diff
@@ -0,0 +1,195 @@
+		value_type operator*() const
+		{
+			// Default to row-major iteration e.g. {0, 0}, {1, 0}, {2, 0}, {0, 1}, {1, 1}, ...
+			return region.position + Displacement { index % region.size.width, index / region.size.width };
```
Not important for small rectangles, but if we start using this in larger rectangles, it'd be nice to avoid a modulo and an integer division on every iteration (can simply keep 2 indices instead of 1).

glebm reviewed (dismissed):
Approval for the C++ iterator stuff

glebm commented on Source/engine/points_in_rectangle_range.hpp:
```diff
@@ -0,0 +1,355 @@
+	using pointer = const value_type *;
+	using reference = value_type &&;
+
+	virtual reference operator*() const = 0;
```
Let's avoid virtual methods, they're slower to call. Simply remove this line, remove `override` throughout.

ephphatha replied:
I've split this into a separate branch for now while I try work out the syntax required to get it building on GCC. The only reason I had this virtual method was to make the base class abstract, but that's not necessary given the base class doesn't fully define any of the required concepts for an iterator.

ephphatha commented:
Finally got around to fixing this on GCC. MSVC appears to use inheritance for iterators with common logic but from what I can see GCC really does not like this approach. There are also footguns with certain operators which return a copy, they need to be redefined to ensure that derived class don't revert back to the base class when those operators are used. I ended up just splitting the definitions and redefining all operators so that there was no chance for one iterator/range to morph into the other.

There are additional uses I haven't included in this PR (`trm5x/y` in `TFit_Obj5`, loops over x/y like `DoUnlight`) since I didn't want to make it too annoying to review, if this is an acceptable approach I'll use it more often throughout the code.

AJenbo commented on Source/objects.cpp:
```diff
@@ -4822,16 +4823,19 @@ void OperateTrap(Object &trap)
+
+		PointsInRectangleRange searchArea { Rectangle { target, 1 } };
+		// look for a player near the trigger (using a reverse search to match vanilla behaviour)
+		auto foundPosition = std::find_if(searchArea.crbegin(), searchArea.crend(), [](Point testPosition) { return InDungeonBounds(testPosition) && dPlayer[testPosition.x][testPosition.y] != 0; });
```
Would it make sens to have `InDungeonBounds()` be part of `PointsInRectangleRange`?

ephphatha replied:
I don't think that would be appropriate for this iterator, it's more concerned with providing a way to access all discrete Points that are contained within a Rectangle. If we wanted to restrict creating rectangles to only in bounds areas then that'd be best enforced in the Rectangle class (but I wouldn't want to do that either). For this case, I'd rather use a wrapper `IsPlayerAtPosition` similar to the `IsObjectAtPosition` helper that is proposed in #3324

AJenbo commented on Source/controls/plrctrls.cpp:
```diff
@@ -122,61 +123,71 @@ int GetDistanceRanged(Point destination)
+		// As the player can not stand on the edge of the map this is safe from OOB
+		int8_t itemId = dItem[targetPosition.x][targetPosition.y] - 1;
+		if (itemId < 0) {
+			// there shouldn't be any multi-tile objects, but just in case only considering positive indexes here
```
Isn't sarcophagus such an object?

ephphatha replied:
This comment should say "...multi-tile (dropped) items...". No idea why I wrote object here.

AJenbo commented on Source/controls/plrctrls.cpp:
```diff
@@ -122,61 +123,71 @@ int GetDistanceRanged(Point destination)
 		return; // Don't look for objects in town
+	}
 
-	for (int xx = -1; xx < 2; xx++) {
-		for (int yy = -1; yy < 2; yy++) {
-			if (dObject[mx + xx][my + yy] == 0)
```
This feels like the change that makes this iterator pay off :+1:

ephphatha replied:
Haha, and the sad part is I want to define a replacement for this use case (and the first step of dropping an item) which looks at adjacent points relative to a given direction (basically the order in #3530 as an iterator).

ephphatha commented on Source/controls/plrctrls.cpp:
```diff
@@ -122,61 +123,71 @@ int GetDistanceRanged(Point destination)
+		// As the player can not stand on the edge of the map this is safe from OOB
+		int8_t itemId = dItem[targetPosition.x][targetPosition.y] - 1;
+		if (itemId < 0) {
+			// there shouldn't be any multi-tile objects, but just in case only considering positive indexes here
```
```suggestion
			// there shouldn't be any items that occupy multiple ground tiles, but just in case only considering positive indexes here

```

AJenbo commented:
Hm going to need a rebase if you don't mind :)
